language/tools/read-write-set/src/dynamic_analysis.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::{ResourceKey, TypeTag},
+};
+use move_model::{
+    model::{FunctionEnv, GlobalEnv},
+    ty::Type,
+};
+use move_vm_runtime::data_cache::MoveStorage;
+use prover_bytecode::{
+    access_path::{AbsAddr, AccessPath, Offset, Root},
+    access_path_trie::AccessPathTrie,
+    read_write_set_analysis::{Access, ReadWriteSetState},
+};
+use resource_viewer::{AnnotatedMoveValue, MoveValueAnnotator};
+use std::ops::Deref;
+
+/// A read/write set state with no unbound formals or type variables
+#[derive(Debug)]
+pub struct ConcretizedFormals(AccessPathTrie<Access>);
+
+/// A read/write set state with no secondary indexes and no unbound formals or type variables
+#[derive(Debug)]
+pub struct ConcretizedSecondaryIndexes(ConcretizedFormals);
+
+impl ConcretizedFormals {
```

Comment:
I was wondering for the concretize step can we carve this section out from the move-model dependency: the reason is that we will need concretize step in the production in order to get an approximation. It would be nice if the inferer can return an abstract structure and have a separate crate to materialize it into an access path.

Reply:
That can definitely be done, although it will be a fair amount of work. I think the steps would be something like:
- Refactor access_path.rs and access_path_trie.rs to not use `move_model::Type`, `move_model::FunctionEnv`, or `move_model::TempIndex`. It might be easier just to copy the files + define a way to convert from the original to the new, but obviously refactoring to avoid code duplication would be preferred if it's feasible
- Change dynamic_analysis.rs to resolve type/function info from the blockchain rather than from move_model

The big question in my mind is whether's it's easier to take those steps or just refactor the whole analysis to work directly on the bytecode without move-model (which we would need if we ever want to use this at runtime). I think the former, but am not sure...